Description: Use the system provided compression library of zlib
Author: Thomas Gaugler <thomas@dadie.net>

diff -ur nsis-2.46.orig/Source/czlib.h nsis-2.46-src/Source/czlib.h
--- nsis-2.46.orig/Source/czlib.h	2009-02-01 15:44:30.000000000 +0100
+++ nsis-2.46-src/Source/czlib.h	2009-12-15 20:40:20.000000000 +0100
@@ -18,7 +18,19 @@
 #define __CZLIB_H__
 
 #include "compressor.h"
-#include "zlib/ZLIB.H"
+#ifdef _WIN32
+#  include "zlib/ZLIB.H"
+#  define deflateInit2(strm, level, method, windowBits, memLevel, strategy) \
+     deflateInit2_(strm, level, method, windowBits, memLevel, strategy, \
+     NULL, sizeof(z_stream))
+#  define streamInit(stream)
+#else
+#  include <zlib.h>
+#  define streamInit(stream) \
+      stream->zalloc = (alloc_func)Z_NULL; \
+      stream->zfree = (free_func)Z_NULL; \
+      stream->opaque = (voidpf)Z_NULL;
+#endif
 
 class CZlib : public ICompressor {
   public:
@@ -27,7 +39,9 @@
     int Init(int level, unsigned int dict_size) {
       stream = new z_stream;
       if (!stream) return Z_MEM_ERROR;
-      return deflateInit(stream, level);
+      streamInit(stream);
+      return deflateInit2(stream, level,
+         Z_DEFLATED, -MAX_WBITS, MAX_MEM_LEVEL, Z_DEFAULT_STRATEGY);
     }
 
     int End() {
diff -ur nsis-2.46.orig/Source/SConscript nsis-2.46-src/Source/SConscript
--- nsis-2.46.orig/Source/SConscript	2007-10-03 15:31:56.000000000 +0200
+++ nsis-2.46-src/Source/SConscript	2009-12-15 20:44:14.000000000 +0100
@@ -66,6 +66,8 @@
 	# XXX will cause problems if makensis is cross compiled
 	# on freebsd, libversion.a exists and gives trouble if linked
 	libs += ['version']
+else:
+	libs += ['z']
 
 AddAvailableLibs(env, libs)
 
@@ -87,7 +89,9 @@
 
 ##### Compile makensis
 
-files = makensis_files + bzip2_files + lzma_files + zlib_files
+files = makensis_files + bzip2_files + lzma_files
+if env['PLATFORM'] == 'win32':
+	files += zlib_files
 
 makensis = env.Program(target, files)
 
diff -ur nsis-2.46.orig/Source/Tests/SConscript nsis-2.46-src/Source/Tests/SConscript
--- nsis-2.46.orig/Source/Tests/SConscript	2009-02-06 15:15:39.000000000 +0100
+++ nsis-2.46-src/Source/Tests/SConscript	2009-12-15 20:44:14.000000000 +0100
@@ -1,3 +1,5 @@
+Import('env AddAvailableLibs')
+
 target = 'test'
 
 tests = Split("""
@@ -65,7 +69,9 @@
 	zlib/INFBLOCK.C
 """)
 
-required += zlib_files
+if env['PLATFORM'] == 'win32':
+	required += zlib_files
+
 required_exehead += zlib_exehead_files
 
 cppunitlibs = Split("""
@@ -87,8 +93,6 @@
 	winver.nsi
 """)
 
-Import('env AddAvailableLibs')
-
 # Test scripts
 env.TestScript(scripts)
 
@@ -97,6 +101,8 @@
 	# XXX will cause problems if tests are cross compiled
 	# on freebsd, libversion.a exists and gives trouble if linked
 	extralibs += ['version']
+else:
+	extralibs += ['z']
 
 AddAvailableLibs(env, extralibs)
 
